src/tracers/tracers.cpp

```diff
@@ -0,0 +1,186 @@
+// publicly, and to permit others to do so.
+
+#include "tracers.hpp"
+#include "../main.hpp"
+#include "utils/error_checking.hpp"
+
+namespace Tracers {
+using namespace parthenon::package::prelude;
+
+std::shared_ptr<StateDescriptor> Initialize(ParameterInput *pin) {
+  auto tracer_pkg = std::make_shared<StateDescriptor>("tracers");
+  const bool enabled = pin->GetOrAddBoolean("tracers", "enabled", false);
+  tracer_pkg->AddParam<bool>("enabled", enabled);
+  if (!enabled) return tracer_pkg;
+
+  Params &params = tracer_pkg->AllParams();
+
+  const int num_tracers = pin->GetOrAddInteger("tracers", "num_tracers", 0);
+  params.Add("num_tracers", num_tracers);
+
+  // Initialize random number generator pool
+  int rng_seed = pin->GetOrAddInteger("tracers", "rng_seed", time(NULL));
+  tracer_pkg->AddParam<>("rng_seed", rng_seed);
+  RNGPool rng_pool(rng_seed);
+  tracer_pkg->AddParam<>("rng_pool", rng_pool);
+
+  // Add swarm of tracers
+  std::string swarm_name = "tracers";
+  Metadata swarm_metadata({Metadata::Provides, Metadata::None});
+  tracer_pkg->AddSwarm(swarm_name, swarm_metadata);
+  Metadata real_swarmvalue_metadata({Metadata::Real});
+  tracer_pkg->AddSwarmValue("id", swarm_name, Metadata({Metadata::Integer}));
+
+  // TODO(pgrete) Add CheckDesired/required for vars
+  // thermo variables
+  tracer_pkg->AddSwarmValue("rho", swarm_name, real_swarmvalue_metadata);
+  tracer_pkg->AddSwarmValue("pressure", swarm_name, real_swarmvalue_metadata);
+  tracer_pkg->AddSwarmValue("vel_x", swarm_name, real_swarmvalue_metadata);
+  tracer_pkg->AddSwarmValue("vel_y", swarm_name, real_swarmvalue_metadata);
+  // TODO(pgrete) check proper handling of <3D sims
+  tracer_pkg->AddSwarmValue("vel_z", swarm_name, real_swarmvalue_metadata);
+
+  // TODO(pgrete) this should be safe because we call this package init after the hydro
+  // one, but we should check if there's direct way to access Params of other packages.
+  const bool mhd = pin->GetString("hydro", "fluid") == "glmmhd";
+
+  PARTHENON_REQUIRE_THROWS(pin->GetString("parthenon/mesh", "refinement") == "none",
+                           "Tracers/swarms currently only supported on uniform meshes.");
+
+  if (mhd) {
+    tracer_pkg->AddSwarmValue("B_x", swarm_name, real_swarmvalue_metadata);
+    tracer_pkg->AddSwarmValue("B_y", swarm_name, real_swarmvalue_metadata);
+    tracer_pkg->AddSwarmValue("B_z", swarm_name, real_swarmvalue_metadata);
+  }
+
+  return tracer_pkg;
+} // Initialize
+
+TaskStatus AdvectTracers(MeshBlockData<Real> *mbd, const Real dt) {
+  auto *pmb = mbd->GetParentPointer();
+  auto &sd = pmb->swarm_data.Get();
+  auto &swarm = sd->Get("tracers");
+
+  const auto ndim = pmb->pmy_mesh->ndim;
+
+  auto &x = swarm->Get<Real>("x").Get();
+  auto &y = swarm->Get<Real>("y").Get();
+  auto &z = swarm->Get<Real>("z").Get();
+
+  auto swarm_d = swarm->GetDeviceContext();
+
+  const auto &prim_pack = mbd->PackVariables(std::vector<std::string>{"prim"});
+
+  // update loop. RK2
```

Comment:
I think leapfrog should give much better accuracy for particle trajectories than RK2. It can be implemented (at second order accuracy) by doing a update that looks like forward Euler, but with velocities averaged over the timestep.